news/_posts/2020-09-17-bc-break-part-1.md

```diff
@@ -0,0 +1,157 @@
+class Money
+{
+	/** @var float */
+	private $amount;
+	/** @var Currency */
+	private $currency;
+
+	public function __construct(float $amount, Currency $currency)
+	{
+		$this->amount = $amount;
+		$this->currency = $currency;
+	}
+```
+
+If a project integrates this library as a dependency and wishes to extend this behavior, a very standard way to do so is to use [inheritance](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)).
+
+For example maybe project A needs a Money object with an identifier ? So it could do:
+
+```php
+class ExtendedMoney extends Money
+{
+	/** string */
+    public $identifier;
+}
+```
+
+This is possible because the class is not [final](https://www.php.net/manual/en/language.oop5.final.php) so it is open for inheritance. It can be then considered that this class, and all of its protected and public properties and methods, are part of the API interface of this project. This means than any non-backwards compatible change to this class is a Breaking Change.
+
+So this is a Breaking Change:
+```php
+class Money
+{
+	/** @var float */
+	private $amount;
+	/** @var Currency */
+	private $currency;
+
+	public function __construct(float $amount, Currency $currency, bool $aNewVariable)
+```
+Changing the signature of the `__construct()` is going to break how the project A uses this library. Because when instantiated, the `ExtendedClass` with the following error message:
+
+```
+PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function __construct(), 2 passed and exactly 3 expected
+```
+
+But this is not a Breaking Change:
+```php
+class Money
+{
+	/** @var float */
+	private $amount;
+	/** @var Currency */
+	private $currency;
+
+	public function __construct(float $amount, Currency $currency, bool $aNewVariable = null)
+```
+By providing a default value for the new argument, previous constructions of the class Money using only two arguments will continue functioning.
+
+## SemVer
+
+In order to make it easier for these developers to predict whether a new version introduces Backward Compatibility Breaks, [Tom Preston Werner](http://tom.preston-werner.com/) has invented a convention: [SemVer](https://semver.org/). [SemVer](https://semver.org/) aims to help people immediately understand, when seeing a new software version, what to expect regarding its compatibility with current software.
+
+SemVer requires a version number to be constituted of 3 parts: MAJOR.MINOR.PATCH (example: 3.5.27) .
+
+### SemVer summary
+
+```
+Given a version number MAJOR.MINOR.PATCH, increment the:
+MAJOR version when you make incompatible API changes,
+MINOR version when you add functionality in a backwards compatible manner, and
+PATCH version when you make backwards compatible bug fixes.
+```
+
+When you release a new version of your software, if the new version only contains bug fixes, it is a patch version and you increase the PATCH number. Theoretically no BC Breaks are to be found in this version.
```

Comment:
Theorically ?

Reply:
Sometimes the issue being fixed inside the patch requires a BC break. For example security issues.